vault/core_metrics.go

```diff
@@ -0,0 +1,112 @@
+package vault
+
+import (
+	"context"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/vault/helper/metricsutil"
+	"github.com/hashicorp/vault/helper/namespace"
+	"github.com/hashicorp/vault/sdk/logical"
+)
+
+// TODO: move emitMetrics into this file.
+
+type kvMount struct {
+	Namespace  *namespace.Namespace
+	MountPoint string
+	Version    string
+	NumSecrets int
+}
+
+func (c *Core) findKvMounts() []*kvMount {
+	mounts := make([]*kvMount, 0)
+
+	c.mountsLock.RLock()
+	defer c.mountsLock.RUnlock()
+
+	for _, entry := range c.mounts.Entries {
+		if entry.Type == "kv" {
+			version, ok := entry.Options["version"]
+			if !ok {
+				version = "1"
+			}
+			mounts = append(mounts, &kvMount{
+				Namespace:  entry.namespace,
+				MountPoint: entry.Path,
+				Version:    version,
+				NumSecrets: 0,
+			})
+		}
+	}
+	return mounts
+}
+
+func (c *Core) kvSecretGaugeCollector(ctx context.Context) ([]metricsutil.GaugeLabelValues, error) {
+	// Find all KV mounts
+	mounts := c.findKvMounts()
+	results := make([]metricsutil.GaugeLabelValues, len(mounts))
+
+	// Context must have root namespace
+	ctx = namespace.RootContext(ctx)
+
+	// Route list requests to all the identified mounts.
+	// (All of these will show up as activity in the vault.route metric.)
+	// Then we have to explore each subdirectory.
+	for i, m := range mounts {
+		results[i].Labels = []metrics.Label{
+			metricsutil.NamespaceLabel(m.Namespace),
+			{"mount_point", m.MountPoint},
+		}
+
+		var subdirectories []string
+		if m.Version == "1" {
+			subdirectories = []string{m.MountPoint}
+		} else {
+			subdirectories = []string{m.MountPoint + "metadata/"}
+		}
+
+		for len(subdirectories) > 0 {
+			// If collection was cancelled, return an empty array.
+			select {
+			case <-ctx.Done():
+				return []metricsutil.GaugeLabelValues{}, nil
+			default:
+				break
+			}
+
+			currentDirectory := subdirectories[0]
+			subdirectories = subdirectories[1:]
+
+			listRequest := &logical.Request{
+				Operation: logical.ListOperation,
+				Path:      currentDirectory,
+			}
+			resp, err := c.router.Route(ctx, listRequest)
+			if err != nil {
+				c.logger.Error("failed to perform internal KV list", "mount_point", m.MountPoint, "error", err)
```

Comment:
I've said this before in another patch and I don't think you were persuaded, but: when it comes to errors encountered while trying to generate metrics, I think it's best to report those errors using other metrics rather than logs.  That way it's easy for metrics consumers to be aware that there may be issues with the metrics they're seeing.

Reply:
I did add a metric that covers reported errors:
https://github.com/hashicorp/vault/blob/4bb08b7a7ef41b61d41a09a242bad8e0df018e56/helper/metricsutil/gauge_process.go#L182
```
		p.sink.IncrCounterWithLabels([]string{"metrics", "collection", "error"},
			1,
			p.labels)
```

Reply:
I change the error handling so that it increments this counter for each KV mount that fails, rather than reporting a single error at the top level.